src/query/expression/src/types/variant.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::ops::Range;
+
+use crate::property::Domain;
+use crate::types::string::StringColumn;
+use crate::types::string::StringColumnBuilder;
+use crate::types::string::StringIterator;
+use crate::types::ArgType;
+use crate::types::DataType;
+use crate::types::GenericMap;
+use crate::types::ValueType;
+use crate::values::Column;
+use crate::values::Scalar;
+use crate::values::ScalarRef;
+
+/// BSON bytes representation of `{v: null}`.
+pub const DEFAULT_BSON: &[u8] = &[0x08, 0x00, 0x00, 0x00, 0x0A, 0x76, 0x00, 0x00];
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct VariantType;
+
+impl ValueType for VariantType {
+    type Scalar = Vec<u8>;
+    type ScalarRef<'a> = &'a [u8];
```

Comment:
Can we use `RawDocumentBuf` and `RawDocument` here? They are wrappers of `Vec<u8>` and `[u8]`.
```rust
pub struct RawDocumentBuf {
    data: Vec<u8>,
}
pub struct RawDocument {
    data: [u8],
}
```

Reply:
There are two reasons `RawDocument` is not used for `Scalar` here:

1.  I reused `StringColumn` for variant columns, it becomes difficult to wrap the result from `StringColumn` into `RawDocument`.
2. 2, the scalar function can easily wrap inputs into `RawDocument` on the fly.

Reply:
ok